In [78]:
import geopandas as gpd
import pandas as pd

# Extract data from files

In [64]:
jpath = 'data/house1.json' # path for json file
xpath = 'data/house2.xml' # path for xml file

## JSON DATA

In [66]:
j = pd.read_json(jpath) # use read json to parse json file
j.head(2)

property_id        lat         lng       addr_street
0        29654 -37.783257  145.000198    2 Short Street
1        84314 -37.851406  145.003433  7 Pridham Street

## XML DATA

In [67]:
x = pd.read_xml('xpath')

XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1 (<string>, line 1)

### Fix XML File

Syntax error : Start tag expected, '<' not found, line 1, column 1
so need to look in the file for xml format

In [ ]:
with open(xpath,'r') as xfile:
    x = xfile.read()  # all xml
print(x[0:300],
x[-100:]) 

<property>
  <property_id>53857</property_id>
  <lat>-37.83034134</lat>
  <lng>144.9947815</lng>
  <addr_street>86 Balmain Street</addr_street>
</property>
<property>
  <property_id>68409</property_id>
  <lat>-37.856339</lat>
  <lng>145.02240700000004</lng>
  <addr_street>25 Clarendon Street</addr_s t>-37.741972</lat>
  <lng>145.018771</lng>
  <addr_street>126 Gower Street</addr_street>
</property>


Do not have wrapper 

In [ ]:
x = "<property_list>" + x + "</property_list>"  # add wrapper

In [79]:
print(x[0:300],
x[-300:])

<property_list><property>
  <property_id>53857</property_id>
  <lat>-37.83034134</lat>
  <lng>144.9947815</lng>
  <addr_street>86 Balmain Street</addr_street>
</property>
<property>
  <property_id>68409</property_id>
  <lat>-37.856339</lat>
  <lng>145.02240700000004</lng>
  <addr_street>25 Clarendon 2265</property_id>
  <lat>-37.983546000000004</lat>
  <lng>145.055963</lng>
  <addr_street>39 Mundy Street</addr_street>
</property>
<property>
  <property_id>31804</property_id>
  <lat>-37.741972</lat>
  <lng>145.018771</lng>
  <addr_street>126 Gower Street</addr_street>
</property></property_list>


In [71]:
x = pd.read_xml(x)

XMLSyntaxError: xmlParseEntityRef: no name, line 3107, column 18 (<string>, line 3107)


Found syntax error: XMLSyntaxError: xmlParseEntityRef: no name, line 3107, column 18
Check line 3107 (index 3106)


In [72]:
with open(xpath, "r") as xl: # open the file again
    alllines = xl.readlines() # use realines to make each line into element of list


In [73]:
alllines[3106]

'  <addr_street>2&3/17-19 Browning Street</addr_street>\n'

< and & are strictly illegal in XML need to change to pre-defined entity
There are 5 pre-defined entity references in XML:

&lt;	<	less than
&gt;	>	greater than
&amp;	&	ampersand 
&apos;	'	apostrophe
&quot;	"	quotation mark

# ref : https://www.w3schools.com/xml/xml_syntax.asp

In [76]:
x = x.replace("&","&amp")

In [77]:
x = pd.read_xml(x)

XMLSyntaxError: EntityRef: expecting ';', line 3107, column 22 (<string>, line 3107)